src/state-mate.ts

```diff
@@ -421,22 +560,230 @@ async function checkNetworkSection(section: NetworkSection, sectionTitle: string
+          resolve(data);
+        } else {
+          reject(new Error(`HTTP status code ${response.statusCode}`));
+        }
+      });
+    }).on('error', (err) => {
+      reject(err);
+    });
+  });
+}
+
+function getExplorerApiUrl(explorerHostname: string, address: string, explorerKey?: string) {
+  let url = `https://${explorerHostname}/api?module=contract&action=getsourcecode&address=${address}`;
+  if (explorerKey) {
+    url = `${url}&apikey=${explorerKey}`;
   }
-  const firstArg = argv[2];
-  if (["--help", "-h"].includes(firstArg)) {
-    log(`TODO`);
-    process.exit(0);
+  return url;
+}
+
+async function _loadContractInfoFromModeExplorer(address: string, explorerHostname: string, explorerKey?: string): Promise<ContractInfoFromExplorer> {
+  const sourcesUrl = getExplorerApiUrl(explorerHostname, address, explorerKey);
+  const sourcesResponse = JSON.parse(await httpGetAsync(sourcesUrl));
+  const contractInfo = sourcesResponse.result[0];
+  const contractName = contractInfo["ContractName"];
+  const abi = JSON.parse(contractInfo["ABI"]) as Abi;
+  let implementation = null;
+  if (contractInfo["IsProxy"] === "true") {
+    const implementationAddress = contractInfo["ImplementationAddress"];
+    assert(implementationAddress);
+    implementation = await _loadContractInfoFromModeExplorer(implementationAddress, explorerHostname, explorerKey);
+  }
+  return { contractName, abi, address, implementation };
+}
+
+async function _loadContractInfoFromEtherscanExplorer(address: string, explorerHostname: string, explorerKey?: string): Promise<ContractInfoFromExplorer> {
+  const sourcesUrl = getExplorerApiUrl(explorerHostname, address, explorerKey);
+  let sourcesResponse = JSON.parse(await httpGetAsync(sourcesUrl));
+  if (sourcesResponse.message.indexOf("rate limit") > -1) {
+    log(`Reached rate limit ${explorerHostname}, waiting for 5 seconds...`);
+    await sleep(5000);
+    sourcesResponse = JSON.parse(await httpGetAsync(sourcesUrl) as string);
+  }
+  if (sourcesResponse.message != "OK") {
+    // TODO: error, but check ContractName instead
+    logErrorAndExit(`Failed to download contract info from ${explorerHostname}: ${sourcesResponse.message}\n${JSON.stringify(sourcesResponse, null, 2)}`);
+  }
+  const contractInfo = sourcesResponse.result[0];
+  const abi = JSON.parse(contractInfo["ABI"]) as Abi;
+  let implementation = null;
+  const implementationAddress = contractInfo["Implementation"];
+  if (implementationAddress) {
+    implementation = await _loadContractInfoFromEtherscanExplorer(implementationAddress, explorerHostname, explorerKey);
+  }
+
+  return { contractName: contractInfo["ContractName"], abi, address, implementation };
+}
+
+async function loadContractInfoFromExplorer(address: string, explorerHostname: string, explorerKey?: string): Promise<ContractInfoFromExplorer> {
+  if (explorerHostname.indexOf("mode.network") > -1) {
```

Comment:
Why are they differently threated? It looks like both use the Etherscan-based. Do we need a separate check for BSC?

Reply:
Unfortunately, part of them are slightly different, the same problem is illustrated in diffyscan: https://github.com/lidofinance/diffyscan/blob/525cc6319515791a1f32906b33f710471ff34506/diffyscan/utils/explorer.py#L80-L87

Reply:
Wow... didn't expect that. Thanks, will know now!